In [14]:
import csv
import os
import math

In [15]:
uniname = {}
with open('./dictionaries/uniname.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        uniname[row[0]]=row[1]
uniname

{'1545931': 'Wirtschaftsuniversität Wien',
 '1546113': 'Universität für Musik und darstellende Kunst Wien',
 '1546123': 'Universität für angewandte Kunst',
 '1546137': 'Medizinische Universität Wien',
 '1546148': 'Universität für Bodenkultur Wien',
 '1546202': 'Technische Universität Wien',
 '1546253': 'Campus der Universität Wien',
 '1546266': 'Webster University Vienna',
 '1546291': 'Universität Wien Hauptgebäude',
 '1546292': 'Universitätssportzentrum Wien'}

In [16]:
unigps = {}
with open('./dictionaries/unigps.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        unigps[row[0]]=row[1]
#unigps

In [17]:
unigpsloc = {}
for k, v in uniname.items():
    if k in unigps.keys():
        unigpsloc[k] = unigps.get(k)
unigpsloc

{'1545931': '48.21384341856702,16.408860446502068',
 '1546113': '48.196989050609844,16.38231295437089',
 '1546123': '48.20838710235376,16.38293943118496',
 '1546137': '48.21963565071546,16.351735473671383',
 '1546148': '48.23654607282594,16.337577480698467',
 '1546202': '48.198976375914846,16.369895225486264',
 '1546253': '48.21624134156441,16.351350792743386',
 '1546266': '48.223429445296574,16.423918403848166',
 '1546291': '48.213012828415096,16.360851928870705',
 '1546292': '48.2054889397581,16.317331757786384'}

## Write Unigpsloc Dict

In [18]:
# Write Unigpsloc Dict
with open('./dictionaries/unigpsloc.csv', 'w', newline='') as csvfile:
    writer = csv.writer(csvfile, delimiter=',')
    for k, v in unigpsloc.items():
        writer.writerow( [ k, v ] )
unigpsloc

{'1545931': '48.21384341856702,16.408860446502068',
 '1546113': '48.196989050609844,16.38231295437089',
 '1546123': '48.20838710235376,16.38293943118496',
 '1546137': '48.21963565071546,16.351735473671383',
 '1546148': '48.23654607282594,16.337577480698467',
 '1546202': '48.198976375914846,16.369895225486264',
 '1546253': '48.21624134156441,16.351350792743386',
 '1546266': '48.223429445296574,16.423918403848166',
 '1546291': '48.213012828415096,16.360851928870705',
 '1546292': '48.2054889397581,16.317331757786384'}

In [19]:
cargps = {}
with open('./dictionaries/cargps.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        cargps[row[0]]=row[1]
#cargps

In [20]:
bikegps = {}
with open('./dictionaries/bikegps.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        bikegps[row[0]]=row[1]
#bikegps

# Calculate Distances

Create dictionaries for all Unis, which map ID of car/bike with distance to respective uni.

    disto[15446266]
        [4166:511]
        [4223:772]
    disto[1546292]
        

## legacy 
Attempted to create Dictioniaries in a Loop, failed in doing so - cannot create/access dictionaries from function results. 


print(unigpsloc.keys())

for k in unigpsloc.keys():
    str("disto"+k) = {}
    
disto1546266 = {}
disto1546253 = {}
disto1546292 = {}
disto1546291 = {}
disto1546113 = {}
disto1545931 = {}
disto1546123 = {}
disto1546137 = {}
disto1546202 = {}
disto1546148 = {}

In [21]:
from math import radians, cos, sin, asin, sqrt
def haversine(lon1, lat1, lon2, lat2):
    """
    Calculate the great circle distance between two points 
    on the earth (specified in decimal degrees)
    """
    # convert decimal degrees to radians 
    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
    # haversine formula 
    dlon = lon2 - lon1 
    dlat = lat2 - lat1 
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * asin(sqrt(a)) 
    km = 6367 * c
    return km

# Calculations

## legacy - tried doing it for each one since creation in Loop failed.
'''
disto1545931 = {}
gpsx=float(unigpsloc.get('1545931').split(",")[0])
gpsy=float(unigpsloc.get('1545931').split(",")[1])
for k, v in cargps.items():
    lon2=float(v.split(",")[0])
    lat2=float(v.split(",")[1])
    dis = haversine(gpsx, gpsy, lon2, lat2)
    if dis < 2.0:
        disto1545931[k] = dis
for k, v in bikegps.items():
    lon2=float(v.split(",")[0])
    lat2=float(v.split(",")[1])
    dis = haversine(gpsx, gpsy, lon2, lat2)
    if dis < 2.0:
        disto1545931[k] = dis
    
disto1545931  
'''


## Main
Create a Dictionary in a Dictionaries, calculate Distance from Universities [first dict] to car and bike gps locations, store all Entries below a certain Radius (here 2km) in the second Dict.
This means the second Dictionary in the University Identifier Dict contains car/bike ID's and distance to Uni (in whichever second Dict it is in)
BikeID always starts with 11, Car with 4 - so we can sort by that afterwards.

In [42]:
disto={}
for unikey in unigpsloc.keys():
    disto[unikey]={} 
    
    gpsx=float(unigpsloc.get(unikey).split(",")[0])
    gpsy=float(unigpsloc.get(unikey).split(",")[1])
    
    for k, v in cargps.items():
        lon2=float(v.split(",")[0])
        lat2=float(v.split(",")[1])
        dis = haversine(gpsx, gpsy, lon2, lat2)
        if dis < 2.0:
            disto[unikey][k] = dis
    for k, v in bikegps.items():
        lon2=float(v.split(",")[0])
        lat2=float(v.split(",")[1])
        dis = haversine(gpsx, gpsy, lon2, lat2)
        if dis < 2.0:
            disto[unikey][k] = dis


In [43]:
for uni in disto.keys():
    print("Plätze: " +str(len(disto[str(uni)]))+ "\t" + uniname[uni])

Plätze: 9	Wirtschaftsuniversität Wien
Plätze: 56	Universität Wien Hauptgebäude
Plätze: 58	Campus der Universität Wien
Plätze: 54	Medizinische Universität Wien
Plätze: 21	Universitätssportzentrum Wien
Plätze: 15	Universität für Bodenkultur Wien
Plätze: 39	Universität für Musik und darstellende Kunst Wien
Plätze: 4	Webster University Vienna
Plätze: 40	Universität für angewandte Kunst
Plätze: 50	Technische Universität Wien


## Adding Statistics
Adding up Car / Bike Totals, Calculating average Distance

st\_ represents statistics_ to filter by afterwards

In [51]:
for unikey in disto.keys():
    car, bike = 0, 0
    carsum, bikesum = 0, 0
    for k, v in disto[str(unikey)].items():
        if k[0] == '4':       #car
            car += 1      
            carsum += float(v)
        if k[0:2] == '11':    #bike
            bike += 1     
            bikesum += float(v)
    caravg = carsum/car
    bikeavg = bikesum/bike
    #print("Cars: " + str(car) + "  \tBikes: " + str(bike) + "  \tAvgDist: " + str(avg)[0:5] +" km\t\t"+ uniname[unikey])
    print("Cars: " + str(car) +"  \t~"+ str(caravg)[0:5]+" km  \tBikes: " + str(bike) +"  \t~"+ str(bikeavg)[0:5] +" km\t\t"+ uniname[unikey])
    
    disto[unikey]['st_car']     = str(car)
    disto[unikey]['st_bike']    = str(bike)
    disto[unikey]['st_caravg']  = str(caravg)
    disto[unikey]['st_bikeavg'] = str(bikeavg)
    disto[unikey]['st_total']   = str(bike + car)

Cars: 2  	~1.063 km  	Bikes: 7  	~1.102 km		Wirtschaftsuniversität Wien
Cars: 22  	~1.347 km  	Bikes: 34  	~1.262 km		Universität Wien Hauptgebäude
Cars: 24  	~1.274 km  	Bikes: 34  	~1.316 km		Campus der Universität Wien
Cars: 20  	~1.229 km  	Bikes: 34  	~1.317 km		Medizinische Universität Wien
Cars: 5  	~1.174 km  	Bikes: 16  	~1.409 km		Universitätssportzentrum Wien
Cars: 5  	~1.400 km  	Bikes: 10  	~1.522 km		Universität für Bodenkultur Wien
Cars: 15  	~1.319 km  	Bikes: 24  	~1.301 km		Universität für Musik und darstellende Kunst Wien
Cars: 2  	~1.491 km  	Bikes: 2  	~1.447 km		Webster University Vienna
Cars: 13  	~1.100 km  	Bikes: 27  	~1.185 km		Universität für angewandte Kunst
Cars: 18  	~1.348 km  	Bikes: 32  	~1.320 km		Technische Universität Wien


In [45]:
for k, v in disto['1546291'].items():
    if k[0:3] == 'st_':
        print(k, v)

st_car 22
st_bike 34
st_bikeavg 1.2627057390234353
st_caravg 1.3478353767037192
st_total 56


In [46]:
for k, v in disto['1546291'].items():      
    if k[0:3] != 'st_':
        #continue 
        print(str(k) +"\t" + str(v))

4201	1.6590645177952748
11253	1.1384145386242672
11213	0.11304465707983724
4200	1.1891390051651736
11270	1.268671303482165
11261	1.7039475487640634
11267	1.3981327407930872
11203	1.37055789044108
4196	0.12932098549392465
11240	1.632638343950048
11258	0.8950766838918074
11271	1.1463466468673345
11251	1.899597335331245
4211	1.7633498641998628
11259	1.8848662372352765
4236	0.4690715208672136
4208	1.424420512108778
11208	0.33720125935904655
11237	1.9077560201016317
11207	1.4630732538220212
4205	1.0915829321347423
11212	1.5417168899635567
11262	0.8661144789722911
11266	1.5524718718882349
4174	1.862051481706106
4206	1.8786954591738438
4219	1.6449941568902697
4214	1.0573053793648928
4197	1.4645964363912516
4234	1.5712605601351253
4198	1.684720693306316
11202	0.5969591277810316
11265	0.6747954543083463
11260	1.618677016920643
4235	1.9392597313146223
11269	1.4491388802837692
11201	0.7365912241406238
11263	0.8964277074938065
11242	1.8171981407510258
11255	1.5174521224886675
4215	1.86570972085276